File: packages/ios/Sources/CivUI/CivFormState.swift
// CivUI — CivFormState for SwiftUI
// Centralized form state management with field registration, validation, and reset.
// Mirrors the web CivFormElement's form participation pattern.

import SwiftUI

/// Centralized form state manager for CivUI forms.
///
/// Components register themselves when `formState` and `name` are provided.
/// Call `validate()` on submit to run required checks and custom validators.
///
/// Usage:
/// ```swift
/// struct MyForm: View {
///     @StateObject private var formState = CivFormState()
///     @State private var name = ""
///     @State private var nameError = ""
///
///     var body: some View {
///         CivForm(errors: $formState.errors, state: formState) {
///             CivTextInput(
///                 label: "Full name",
///                 value: $name,
///                 error: nameError.isEmpty ? nil : nameError,
///                 isRequired: true,
///                 formState: formState,
///                 formName: "name"
///             )
///             Button("Submit") {
///                 if formState.validate() {
///                     // submit
///                 }
///             }
///         }
///     }
/// }
/// ```
public class CivFormState: ObservableObject {
    @Published public var errors: [CivFormFieldError] = []
    private var fields: [String: CivFieldRegistration] = [:]

    public init() {}

    /// Registration info for a form-participating field.
    public struct CivFieldRegistration {
        public let name: String
        public let label: String
        public let getValue: () -> String
        public let setValue: (String) -> Void
        public let isRequired: Bool
        public let requiredMessage: String
        public let getError: () -> String
        public let setError: (String) -> Void
        public let validate: (() -> String?)?
        public let isPii: Bool

        public init(
            name: String,
            label: String = "",
            getValue: @escaping () -> String,
            setValue: @escaping (String) -> Void,
            isRequired: Bool = false,
            requiredMessage: String = "",
            getError: @escaping () -> String = { "" },
            setError: @escaping (String) -> Void = { _ in },
            validate: (() -> String?)? = nil,
            isPii: Bool = false
        ) {
            self.name = name
            self.label = label
            self.getValue = getValue
            self.setValue = setValue
            self.isRequired = isRequired
            self.requiredMessage = requiredMessage
            self.getError = getError
            self.setError = setError
            self.validate = validate
            self.isPii = isPii
        }
    }

    /// Register a field for form state management.
    public func register(_ reg: CivFieldRegistration) {
        fields[reg.name] = reg
    }

    /// Unregister a field (typically called in `.onDisappear`).
    public func unregister(_ name: String) {
        fields.removeValue(forKey: name)
    }

    /// Validate all registered fields. Returns `true` if no errors.
    ///
    /// Runs required checks first, then custom validators for non-empty fields.
    /// Sets error messages on individual fields and populates the `errors` array.
    @discardableResult
    public func validate() -> Bool {
        var newErrors: [CivFormFieldError] = []
        clearErrors()

        for (_, field) in fields.sorted(by: { $0.key < $1.key }) {
            // Required check
            if field.isRequired && field.getValue().isEmpty {
                let fallbackLabel = field.label.isEmpty ? field.name : field.label
                let msg = field.requiredMessage.isEmpty
                    ? CivLocale.shared.t("fieldRequired").replacingOccurrences(of: "{label}", with: fallbackLabel)
                    : field.requiredMessage
                newErrors.append(CivFormFieldError(fieldName: field.name, message: msg))
                field.setError(msg)
                continue
            }

            // Custom validation
            if let customValidate = field.validate, !field.getValue().isEmpty {
                if let error = customValidate() {
                    newErrors.append(CivFormFieldError(fieldName: field.name, message: error))
                    field.setError(error)
                }
            }
        }

        errors = newErrors
        return errors.isEmpty
    }

    /// Reset all fields to empty values and clear all errors.
    public func reset() {
        errors = []
        fields.values.forEach { $0.setValue(""); $0.setError("") }
    }

    /// Clear all errors without resetting field values.
    public func clearErrors() {
        errors = []
        fields.values.forEach { $0.setError("") }
    }

    /// Get form data as a dictionary, excluding PII fields.
    public func getFormData() -> [String: String] {
        var data: [String: String] = [:]
        for (name, field) in fields {
            if !field.isPii {
                data[name] = field.getValue()
            }
        }
        return data
    }

    /// Get form data including all fields (use with caution for PII).
    public func getAllFormData() -> [String: String] {
        var data: [String: String] = [:]
        for (name, field) in fields {
            data[name] = field.getValue()
        }
        return data
    }
}

File: packages/android/src/main/kotlin/gov/civui/components/CivFormState.kt
// CivUI — CivFormState for Jetpack Compose
// Manages form field registration, validation, reset, and error state.
// Enables declarative form validation patterns for government applications.

package gov.civui.components

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import gov.civui.i18n.CivLocale

/**
 * Form state manager for CivUI Compose forms.
 *
 * Coordinates field registration, validation, reset, and error
 * collection across form components. Fields register themselves
 * via [register] (typically in a DisposableEffect) and unregister
 * when removed from composition.
 *
 * Usage:
 * ```kotlin
 * val formState = remember { CivFormState() }
 *
 * CivForm(
 *     state = formState,
 *     onSubmit = {
 *         if (formState.validate()) {
 *             val data = formState.getFormData()
 *             // submit data
 *         }
 *     },
 * ) { submitHandler ->
 *     CivTextInput(
 *         label = "Full name",
 *         value = name,
 *         onValueChange = { name = it },
 *         required = true,
 *         formState = formState,
 *         name = "fullName",
 *     )
 * }
 * ```
 */
class CivFormState {
    var errors by mutableStateOf<List<CivFormFieldError>>(emptyList())
        private set

    private val fields = mutableMapOf<String, CivFieldRegistration>()

    /**
     * Registration data for a single form field.
     *
     * @param name Unique field name (used as key)
     * @param getValue Returns the current field value
     * @param setValue Sets the field value (for reset)
     * @param isRequired Whether the field is required
     * @param requiredMessage Custom "required" error message
     * @param getError Returns the current error string for this field
     * @param setError Sets the error string for this field
     * @param validate Optional custom validation function; returns error string or null
     * @param isPii If true, field value is excluded from getFormData()
     */
    data class CivFieldRegistration(
        val name: String,
        val getValue: () -> String,
        val setValue: (String) -> Unit,
        val isRequired: Boolean,
        val requiredMessage: String = "",
        val getError: () -> String,
        val setError: (String) -> Unit,
        val validate: (() -> String?)? = null,
        val isPii: Boolean = false,
    )

    /**
     * Register a field with this form state. Called from DisposableEffect.
     */
    fun register(reg: CivFieldRegistration) {
        fields[reg.name] = reg
    }

    /**
     * Unregister a field. Called from DisposableEffect onDispose.
     */
    fun unregister(name: String) {
        fields.remove(name)
    }

    /**
     * Validate all registered fields. Returns true if no errors.
     *
     * Checks required fields first, then runs custom validate functions.
     * Sets error messages on individual fields and populates [errors].
     */
    fun validate(): Boolean {
        clearErrors()
        val newErrors = mutableListOf<CivFormFieldError>()
        fields.values.sortedBy { it.name }.forEach { field ->
            // Check required
            if (field.isRequired && field.getValue().isEmpty()) {
                val msg = field.requiredMessage.ifEmpty {
                    CivLocale.t("fieldRequired", "label" to field.name)
                }
                newErrors.add(CivFormFieldError(field.name, msg))
                field.setError(msg)
                return@forEach
            }
            // Run custom validation on non-empty fields
            if (field.getValue().isNotEmpty()) {
                field.validate?.invoke()?.let { error ->
                    newErrors.add(CivFormFieldError(field.name, error))
                    field.setError(error)
                }
            }
        }
        errors = newErrors
        return errors.isEmpty()
    }

    /**
     * Reset all fields to empty and clear all errors.
     */
    fun reset() {
        errors = emptyList()
        fields.values.forEach {
            it.setValue("")
            it.setError("")
        }
    }

    /**
     * Clear all errors without resetting field values.
     */
    fun clearErrors() {
        errors = emptyList()
        fields.values.forEach { it.setError("") }
    }

    /**
     * Collect form data as a name-value map, excluding PII fields.
     */
    fun getFormData(): Map<String, String> {
        return fields.filter { !it.value.isPii }.mapValues { it.value.getValue() }
    }
}
